File: Data Structures & Algorithms/combination-target-sum/submission-3.cpp
class Solution {
private:
    void solve(int index,vector<vector<int>>& ans,vector<int>& output,vector<int>& nums,int target){
        if(target==0){
            ans.push_back(output);
            return;
        }

        if(target<0 || index==nums.size()){ return;}

        if(nums[index]<=target){
            output.push_back(nums[index]);
            solve(index,ans,output,nums,target-nums[index]);
            output.pop_back();
        }
        solve(index+1,ans,output,nums,target);

    }
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> ans;
        vector<int> output;
        solve(0,ans,output,nums,target);
        return ans;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        nums =sorted(nums)

        for k in range(len(nums)):
            if (k>0) and (nums[k]==nums[k-1]):
                continue

            i=k+1
            j=len(nums)-1

            while(i<j):
                sum= nums[i]+nums[j]+nums[k]

                if(sum==0):
                    ans.append([nums[i],nums[j],nums[k]])
                    i+=1
                    j-=1
                
                    while(i<j and nums[i]==nums[i-1]):
                        i+=1;
                    while(i<j and nums[j]==nums[j+1]):
                        j-=1;


                if sum < 0:
                    i+=1

                elif sum > 0:
                    j-=1

        
        return ans

                